## Production Model

In [637]:
model8_features = ['BsmtFin SF 1','Overall Qual',
 'Total Bsmt SF',
 '1st Flr SF',
 'Gr Liv Area',
 'Mas Vnr Area',
 '2nd Flr SF',
 'Garage Area',
 'paved',
 'new or con',
 'Central Air',
 'Electrical',
 'Kitchen Qual',
'Neighborhood Codes' ]

In [638]:
x8_train, x8_test, y8_train, y8_test = train_test_split(train_data_7[model8_features],train_data_7['SalePrice'],random_state=42)

In [641]:
m8_mlr=LinearRegression()

In [642]:
m8_mlr.fit(x8_train, np.log(y8_train))

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [643]:
m8_train_preds = m8_mlr.predict(x8_train)

In [688]:
model_metrics(y8_train, np.exp(m8_train_preds))

The mean of the residuals is 1253.16
Mean absolute error is 17137.33
The sum squared error is 891560814514.71
The mean squared error is 587325964.77
The root mean squared error is 24234.81
The r2 score is 0.90


In [691]:
m8_test_preds = m8_mlr.predict(x8_test)

In [692]:
model_metrics(y8_test, np.exp(m8_test_preds))

The mean of the residuals is 1395.24
Mean absolute error is 16582.70
The sum squared error is 278563269179.22
The mean squared error is 550520294.82
The root mean squared error is 23463.17
The r2 score is 0.92


In [648]:
m8_mlr.fit(train_data_7[model8_features], np.log(train_data_7['SalePrice']))

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [657]:
m8_preds = m8_mlr.predict(train_data_7[model8_features])

In [650]:
model_metrics(np.log(train_data_7['SalePrice']), m8_preds)

The mean of the residuals is 0.00
Mean absolute error is 0.10
The sum squared error is 42.49
The mean squared error is 0.02
The root mean squared error is 0.14
The r2 score is 0.88


In [679]:
list(zip(model8_features, (np.exp(m8_mlr.coef_)-1)*100))

[('BsmtFin SF 1', 0.01241492617645168),
 ('Overall Qual', 8.752426671074032),
 ('Total Bsmt SF', 0.00821728762019358),
 ('1st Flr SF', 0.02796799717548204),
 ('Gr Liv Area', 0.002466739977857735),
 ('Mas Vnr Area', -0.0025812330096242064),
 ('2nd Flr SF', 0.023019858818051908),
 ('Garage Area', 0.014171586648448198),
 ('paved', 11.401216177488461),
 ('new or con', 7.392199381496223),
 ('Central Air', 16.663164608067383),
 ('Electrical', 2.2904328013664266),
 ('Kitchen Qual', 6.251802312331289),
 ('Neighborhood Codes', 15.153394851547187)]

In [757]:
np.exp(m8_mlr.intercept_)

27473.700138398777

### Hypothesis Testing

In [753]:
hyp_model = sm.OLS(np.log(train_data_7['SalePrice']), train_data_7[model8_features]).fit()

In [754]:
hyp_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:              SalePrice   R-squared (uncentered):                   0.990
Model:                            OLS   Adj. R-squared (uncentered):              0.990
Method:                 Least Squares   F-statistic:                          1.394e+04
Date:                Thu, 09 Apr 2020   Prob (F-statistic):                        0.00
Time:                        16:10:51   Log-Likelihood:                         -3265.3
No. Observations:                2024   AIC:                                      6559.
Df Residuals:                    2010   BIC:                                      6637.
Df Model:                          14                                                  
Covariance Type:            nonrobust                                                  
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
BsmtFin SF 1          -0.0002   7.34e-05     -2.405      0.016      -0.000   -3.25e-05
Overall Qual           0.2689      0.034      7.972      0.000       0.203       0.335
Total Bsmt SF         -0.0007      0.000     -5.780      0.000      -0.001      -0.000
1st Flr SF             0.0016      0.001      2.890      0.004       0.001       0.003
Gr Liv Area            0.0002      0.001      0.453      0.651      -0.001       0.001
Mas Vnr Area          -0.0014      0.000     -7.972      0.000      -0.002      -0.001
2nd Flr SF             0.0002      0.001      0.321      0.748      -0.001       0.001
Garage Area           -0.0010      0.000     -5.795      0.000      -0.001      -0.001
paved                  0.6702      0.104      6.436      0.000       0.466       0.874
new or con            -0.6283      0.111     -5.650      0.000      -0.846      -0.410
Central Air            0.3631      0.124      2.929      0.003       0.120       0.606
Electrical             2.3458      0.056     42.115      0.000       2.237       2.455
Kitchen Qual           0.2962      0.059      5.021      0.000       0.181       0.412
Neighborhood Codes     0.7397      0.117      6.297      0.000       0.509       0.970
==============================================================================
Omnibus:                     1111.623   Durbin-Watson:                   1.893
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9915.613
Skew:                           2.454   Prob(JB):                         0.00
Kurtosis:                      12.668   Cond. No.                     1.16e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.16e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

With 5% significance, it is conclusive to say that general living area and second floor square footage have a linear relationship to the log of sale price. 

## Conclusion and Recommendation

|Predictors|% Change in Sale Price if Predictor is Increased by 1 Unit|
|-|-|
|Central Air|16.66%|
|Neighborhood Codes|15.15%|
|Paved|11.4%|
|Overall Quality|8.75%|
|Sale Type (New or Con)|7.39%|
|Kitchen Quality|6.25%|
|Electrical|2.29%|
|1st Floor Square Foot|0.028%|
|2nd Floor SF|0.023%|
|Garage Area|0.014%|
|Type 1 finished square feet |0.012% |
|Total Basement Square Foot|0.008%|
|Mason Veneer Area|-0.0026%|
|General Living Area|0.0025%|




The above chart highlights the 14 predictors that are used in this model to assess the housing prices in Ames. While it is easy to conclude that the predictor with the highest percentage change in sale price is the strongest influencer, it is important to quantify each predictor before making that conclusion. For example, Central Air is a categorical data that contains two data points: 0 or 1. 0 for houses without central air and 1 for houses with central air. With values that are 0 or 1, it is more likely that the percentage change is inflated that it actually is in real life. Another way to quantify the predictors is to consider the real life application. For example, general living area has the lowest percentage, but the predictors also have the highest range. Let's take a house that is selling for 500,000 dollars, if the general living area's square footage is increased by 1 unit, the sale price is predicted to increase by 500,000 * 1.0025 = 501,250 dollars, with all else hold equal. A typical bedroom size is about 132 square foot so general living area actually has a greater influence than central air. 

Two of the predictors here are especially susceptible to data entry errors or bias during rating: Kitchen Quality and Overall Quality. It is paramount for Ames Assessor's Office to ensure that all their agents understand the guidelines on how to rate each house and standardize the rating rubric to minimize bias. In addition, the office should consider cross validating each agent's scores by sending two or three agents to evaluate the house to prevent bias. The Ames Assessor's Office's evaluation is crucial to the wealth of the city and residents. Taking these precautionary steps to validate the data will ensure that the model perform optimally to assess housing prices as accurate as possible. 

### Kaggle Submission

In [658]:
to_submit4 = pd.DataFrame()
to_submit4['Id'] = test7['Id']
m8_test_preds = m8_mlr.predict(test7[model8_features])


In [659]:
to_submit4['SalePrice'] = np.exp(m8_test_preds)

In [660]:
to_submit4

Id      SalePrice
0    2658  122580.456503
1    2718  175483.219988
2    2414  198826.688008
3    1989  105563.619290
4     625  162368.676726
..    ...            ...
873  1662  197425.523194
874  1234  190654.754146
875  1373  127556.556682
876  1672  112894.920624
877  1939  125988.384451

[878 rows x 2 columns]

In [662]:
to_submit4.to_csv('./datasets/to submit4.csv', index=False)